tree/ntuple/v7/src/RNTupleProcessor.cxx

```diff
@@ -352,135 +349,122 @@ ROOT::Experimental::RNTupleJoinProcessor::RNTupleJoinProcessor(const RNTupleOpen
       // used in the processor loop to access the corresponding field values.
-      if (!models.empty() && !models[0]->IsBare()) {
-         auto valuePtr = models[0]->GetDefaultEntry().GetPtr<void>(fieldName);
+      if (!fModel->IsBare()) {
+         auto valuePtr = fModel->GetDefaultEntry().GetPtr<void>(fieldName);
          fEntry->BindValue(fieldName, valuePtr);
       }
 
-      const auto &[fieldContext, _] =
-         fFieldContexts.try_emplace(fieldName, field.Clone(fieldName), fEntry->GetToken(fieldName));
-      ConnectField(fieldContext->second, *fPageSource, *fEntry);
+      fFieldContexts.try_emplace(fieldName, field.Clone(fieldName), fEntry->GetToken(fieldName));
    }
 
    for (unsigned i = 0; i < auxNTuples.size(); ++i) {
-      if (models.empty()) {
-         AddAuxiliary(auxNTuples[i], joinFields);
-      } else {
-         // The size of `models` is checked in `CreateJoin`; at this point we can safely assume that `models.size() ==
-         // auxNTuples.size() + 1`.
-         AddAuxiliary(auxNTuples[i], joinFields, std::move(models[i + 1]));
-      }
+      AddAuxiliary(auxNTuples[i], joinFields, i + 1 /* ntupleIdx */);
    }
 }
 
-void ROOT::Experimental::RNTupleJoinProcessor::AddAuxiliary(const RNTupleOpenSpec &auxNTuple,
-                                                            const std::vector<std::string> &joinFields,
-                                                            std::unique_ptr<RNTupleModel> model)
+void ROOT::Experimental::RNTupleJoinProcessor::SetModel(std::unique_ptr<RNTupleModel> primaryModel,
+                                                        std::vector<std::unique_ptr<RNTupleModel>> auxModels)
 {
-   assert(fNEntriesProcessed == 0 && "cannot add auxiliary ntuples after processing has started");
-
-   fNTuples.emplace_back(auxNTuple);
-
-   auto pageSource = Internal::RPageSource::Create(auxNTuple.fNTupleName, auxNTuple.fStorage);
-   pageSource->Attach();
-
-   if (pageSource->GetNEntries() == 0) {
-      throw RException(R__FAIL("provided RNTuple is empty"));
-   }
-
-   if (!model)
-      model = pageSource->GetSharedDescriptorGuard()->CreateModel();
+   fModel = std::move(primaryModel);
+   fModel->Unfreeze();
 
-   model->Freeze();
-   auto entry = model->CreateBareEntry();
+   // Create an anonymous record field for each auxiliary ntuple, containing their top-level fields. These original
+   // top-level fields are registered as subfields in the join model, such that they can be accessed as
+   // `auxNTupleName.fieldName`.
+   // TODO(fdegeus) Support projected fields
+   for (unsigned i = 0; i < auxModels.size(); ++i) {
+      std::vector<std::unique_ptr<RFieldBase>> auxFields;
 
-   // Append the auxiliary fields to the join model
-   fModel->Unfreeze();
+      if (!auxModels[i]) {
+         fAuxiliaryPageSources[i]->Attach();
+         auxModels[i] = fAuxiliaryPageSources[i]->GetSharedDescriptorGuard()->CreateModel();
+      }
```

Comment:
Maybe this can be moved to the `RNTupleJoinProcessor` constructor? In case of an empty vector, `resize` with `nullptr` and then generally replace all `nullptr` with a reconstructed model?

Reply:
Done, see lines 358-366.